### PR TITLE
Move to S6

### LINKED ISSUE #233: ntfy not found in b3vis/borgmatic:ntfy-1.6.6 and later
When running the latest image `b3vis/borgmatic:ntfy-latest`, the `ntfy` tool is not found:

```bash
$ docker run --rm b3vis/borgmatic:ntfy-latest ntfy
Unable to find image 'b3vis/borgmatic:ntfy-latest' locally
docker: Error response from daemon: manifest for b3vis/borgmatic:ntfy-latest not found: manifest unknown: manifest unknown.
See 'docker run --help'.
```

I've traced this back to the 1.6.5 version - that is the latest version containing a working `ntfy` tool:

```shell
docker run --rm b3vis/borgmatic:ntfy-1.6.5 ntfy
usage: ntfy [-h] [-c CONFIG] [-b BACKEND] [-o key value]
            [-l {CRITICAL,ERROR,WARNING,INFO,DEBUG}] [-v] [-q] [--version]
            [-E] [-t TITLE]
            {send,done,shell-integration} ...

Send push notification when command finishes

positional arguments:
  {send,done,shell-integration}
    send                send a notification
    done                run a command and send a notification when done
    shell-integration   automatically get notifications when long running
                        commands finish

optional arguments:
  -h, --help            show this help message and exit
  -c CONFIG, --config CONFIG
                        config file to use (default:
                        /root/.config/ntfy/ntfy.yml)
  -b BACKEND, --backend BACKEND
                        override backend specified in config
  -o key value, --option key value
                        backend specific options
  -l {CRITICAL,ERROR,WARNING,INFO,DEBUG}, --log-level {CRITICAL,ERROR,WARNING,INFO,DEBUG}
                        Specify the how verbose CLI output is (default:
                        WARNING)
  -v, --verbose         a shortcut for --log-level=DEBUG
  -q, --quiet           a shortcut for --log-level=CRITICAL
  --version             show program's version number and exit
  -E, --no-emoji        Disable emoji support
  -t TITLE, --title TITLE
                        a title for the notification (default:
                        root@27ceeb16f960:/)
```

With 1.6.6:

```shell
$ docker run --rm b3vis/borgmatic:ntfy-1.6.6 ntfy
Traceback (most recent call last):
  File "/usr/bin/ntfy", line 5, in <module>
    from ntfy.cli import main
ModuleNotFoundError: No module named 'ntfy'
```

### LINKED ISSUE #242: [msmtp] Not using MAIL_SUBJECT
Borgmatic Version: 1.7.15

Mails get sent fine but only have `✅ SUCCESS:` or `❌ FAILED:` as subject without the value of `MAIL_SUBJECT`

### PR DESCRIPTION
**Pull request by modem7**
*This is a WIP PR to move us to S6 and simplify + make the image more resilient.*

- Added S6 (which should also fix https://github.com/borgmatic-collective/docker-borgmatic/pull/173)
- Added Docker env var to allow for docker operations within the container
- Added Cron var for additional cron jobs
- Removed dedicated Apprise package as that is now part of the Borgmatic package.
- Removed Supercronic in favour of basic cron
- Entrypoint moved to /root/etc/s6-overlay/s6-rc.d/svc-cron/run (This needs reviewing the most as there has been additional change/reversions that may need resolving/modifying).

# Environment Variables
| Variable | Description | Possible Values | Default |
| :----: | --- | --- | --- |
| DOCKERCLI | Install DockerCLI | true | Empty |
| CRON | Cron times | cron time/false | 0 1 * * * |
| CRON_COMMAND | Command cron will run | borgmatic --stats -v 0 2>&1 | borgmatic --stats -v 0 2>&1 |
| EXTRA_CRON | Extra cron lines | 0 5 2 * * command1 | Empty |

- @Psycho0verload please could you review the changes that will be required to cover the changes that were done in #275. I haven't had time to implemented them yet, and I don't want to introduce a regression.

Once merged, this PR will:
close #233
close #242
close #272
close #301

### CODE REVIEW

**Comment by Psycho0verload** [comment]
I will look at it in the coming days

File: Dockerfile
Comment (grantbevis) [inline Dockerfile:6]: Any chance you could change this to `borgmatic-collective` please dude

Comment (modem7) [reply Dockerfile]: Dun n dun. Also changed the repo in the requirements.txt

File: requirements.txt
Comment (grantbevis) [inline requirements.txt]: Another instance :)

Comment (modem7) [reply requirements.txt]: Both done in https://github.com/borgmatic-collective/docker-borgmatic/pull/303/commits/ea6d942f3588e4b61843f4b9b9e5a81ec048288f

**Review by grantbevis — DISMISSED** [review]
LGTM!

**Comment by grantbevis** [comment]
Will wait for @Psycho0verload 's review too

**Comment by modem7** [comment]
> Will wait for @Psycho0verload 's review too

As soon as @Psycho0verload gives his blessing that we haven't broken his work (once we modify the entrypoint script), we'll get it merged and close the linked PR's/issues.

File: Dockerfile
Comment (witten) [inline Dockerfile:82]: Let me know if it would be helpful to have a dedicated healthcheck of some sort in borgmatic.

File: data/borgmatic.d/config.yaml
Comment (witten) [inline data/borgmatic.d/config.yaml:1]: FYI these section headers are deprecated in recent borgmatic versions.

File: Dockerfile
Comment (modem7) [reply Dockerfile:82]: > Let me know if it would be helpful to have a dedicated healthcheck of some sort in borgmatic.

If we can, that would be great! Not quite sure how one would do it, but if we can figure out a healthcheck which would check for base functionality of "app is at least up and running, and shouldn't have an issue with the current config", it would be fantastic!

File: data/borgmatic.d/config.yaml
Comment (modem7) [reply data/borgmatic.d/config.yaml:1]: > FYI these section headers are deprecated in recent borgmatic versions.

This is what I get for not keeping up to date on the configs myself! Lemme check the relevant config and replace!

Comment (witten) [reply data/borgmatic.d/config.yaml:1]: I'll admit even I had systems with out-of-date borgmatic config files for an embarrassingly long time. :smile:

File: Dockerfile
Comment (modem7) [reply Dockerfile:82]: > I know I just offered to add a healthcheck, but maybe the existing `borgmatic config validate` already does what you're describing? If it doesn't in some way, I'm happy to change or add to it.

In your opinion, would that be a suitable check for the application (and would it output to the general log files)? The latter question is more a case of "we don't want to spam the logs if it's OK, and we want to make sure it exits the container with an appropriate exit code if it's not working as intended"

File: data/borgmatic.d/config.yaml
Comment (modem7) [reply data/borgmatic.d/config.yaml]: @witten - This better?

https://github.com/borgmatic-collective/docker-borgmatic/pull/303/commits/d0953b44680848bc03383f499835d6c22d7bd6c7

File: Dockerfile
Comment (witten) [reply Dockerfile:82]: You could always add `--verbosity -1` (errors only) if you didn't want any output from borgmatic other than errors. And then, yeah, it would probably be a suitable check. It of course won't tell you if any configured remote repositories are reachable or anything like that, but it's a good first step.

File: data/borgmatic.d/config.yaml
Comment (witten) [reply data/borgmatic.d/config.yaml]: Yes, looks great! The only other thing you might get a deprecation warning on is the repositories list. If you change it to this, the warning will go away:

```yaml
repositories:
    - path: /mnt/borg-repository
```

I'd also recommend removing `prefix:` entirely. It's deprecated and `archive_name_format` supersedes it.

Comment (modem7) [reply data/borgmatic.d/config.yaml]: Dun n dun! Shall let you review and close if good!

Comment (witten) [reply data/borgmatic.d/config.yaml]: LGTM!

File: Dockerfile
Comment (modem7) [reply Dockerfile]: Updated healthcheck in https://github.com/borgmatic-collective/docker-borgmatic/pull/303/commits/d3155314b384c7d14b17d63590e61dd705a20942

Have checked that exit code 1 is given when an invalid config is presented, and have also checked that the console does not get filled up even without the `verbosity -1` flag.

Comment (witten) [reply Dockerfile]: Looks good! Just note that without `--verbosity -1`, borgmatic produces output when the configs files are valid. If that's not filling up the console, great.

Comment (modem7) [reply Dockerfile]: Aye, honestly thought it would fill up the console to be fair.

I did a curiosity test as with verbosity -1 it also doesn't give a healthcheck output. So it was kinda win/lose.

With no verbosity flag, it fills the healthcheck output, but not the console logs. So an unexpected win/win there.

Comment (witten) [reply Dockerfile]: Ah gotcha, cool. Makes sense that healthcheck output would be separate.

**Review by grantbevis — DISMISSED** [review]

**Comment by modem7** [comment]
The conflict can be easily resolved. 

@Psycho0verload - did you manage to get some time to have a look at the `/root/etc/s6-overlay/s6-rc.d/svc-cron/run` file?

**Comment by Psycho0verload** [comment]
For your information: I was able to implement my script and it works as far as I could test it. What I came across in the documentation:
```
YOUR_PASSPHRASE
```

I would extend my script for this and the push to the branch will follow today.

**Comment by Psycho0verload** [comment]
@modem7 added a PR on your Fork.

**Comment by modem7** [comment]
I have merged @Psycho0verload's PR (thank you for that), and have resolved the merge conflict. 

Ready for review and final comments prior to merging.

File: data/borgmatic.d/config.yaml
Comment (witten) [inline data/borgmatic.d/config.yaml:8]: Typo here: "Don"

File: root/etc/s6-overlay/s6-rc.d/svc-cron/run
Comment (witten) [inline root/etc/s6-overlay/s6-rc.d/svc-cron/run:84]: This checks for `false`/`disabled`/`off` while `DEBUG_SECRETS` above checks for `true`/`1`. Is this inconsistency intentional or for backwards compatibility maybe?

File: root/etc/s6-overlay/s6-rc.d/svc-cron/run
Comment (witten) [inline root/etc/s6-overlay/s6-rc.d/svc-cron/run:61]: I might be missing something, but how do these environment variable exports make their way into the borgmatic invocation that cron runs?

**Review by witten — COMMENTED** [review]
Cursory review as I'm not familiar with S6 or most of the moving parts here other than borgmatic.

File: root/etc/s6-overlay/s6-rc.d/svc-cron/run
Comment (modem7) [reply root/etc/s6-overlay/s6-rc.d/svc-cron/run]: Resolved in https://github.com/borgmatic-collective/docker-borgmatic/pull/303/commits/43a5d3d3aec0eedd9dfef22cb436d9fdc96bb3e5

Comment (modem7) [reply root/etc/s6-overlay/s6-rc.d/svc-cron/run]: It was mostly a historical thing, but looking at it, there's really no reason to include multiple options.

File: data/borgmatic.d/config.yaml
Comment (modem7) [reply data/borgmatic.d/config.yaml]: Resolved in https://github.com/borgmatic-collective/docker-borgmatic/pull/303/commits/43a5d3d3aec0eedd9dfef22cb436d9fdc96bb3e5

File: root/etc/s6-overlay/s6-rc.d/svc-cron/run
Comment (modem7) [reply root/etc/s6-overlay/s6-rc.d/svc-cron/run:61]: @Psycho0verload - one for you

Comment (Psycho0verload) [reply root/etc/s6-overlay/s6-rc.d/svc-cron/run:61]: @witten is right.

Unfortunately, I am not familiar with S6 either. That's why I found it easiest to integrate the script into the script call, which runs in any case.

I have looked at S6 and probably the best solution would be to make a separate call for the secrets. @modem7 maybe you are more familiar with this and can build your own call for it. The script itself works. 

Otherwise I would experiment with it.

Comment (modem7) [reply root/etc/s6-overlay/s6-rc.d/svc-cron/run:61]: For now, I would say let's keep it simple. If the script works for now, then that's all we need. Once we're happy with how the full PR is, and that it works as intended with no bugs from the migration, we can start investigating S6 in its entirety, and look at expanding it with additional calls.

We'll be a lot less confined with S6, and we can add/remove things in a modular fashion going forwards, making it much easier to test and develop.

Comment (grantbevis) [reply root/etc/s6-overlay/s6-rc.d/svc-cron/run:61]: Are we all happy with this then?

Comment (modem7) [reply root/etc/s6-overlay/s6-rc.d/svc-cron/run:61]: LGTM

Comment (grantbevis) [reply root/etc/s6-overlay/s6-rc.d/svc-cron/run:61]: Will await @Psycho0verload 's review then I'll tick and we can merge this bad boy

**Review by Psycho0verload — APPROVED** [review]

**Comment by modem7** [comment]
Merge conflicts resolved. We are ready to merge.

Once merged, close https://github.com/borgmatic-collective/docker-borgmatic/issues/233 https://github.com/borgmatic-collective/docker-borgmatic/issues/242 https://github.com/borgmatic-collective/docker-borgmatic/pull/272 https://github.com/borgmatic-collective/docker-borgmatic/pull/301

**Review by grantbevis — APPROVED** [review]
🫡

**Comment by rcdailey** [comment]
Hey folks, I think it might be helpful to announce that if people were previously using `init: true` in their docker compose files, they should remove that, otherwise you get this error (not surprising, since the container is using overlay now):

> s6-overlay-suexec: fatal: can only run as pid 1

I used the init setting because before overlay was introduced, there was nothing to properly handle interrupt signals in the container.

**Comment by modem7** [comment]
> Hey folks, I think it might be helpful to announce that if people were previously using `init: true` in their docker compose files, they should remove that, otherwise you get this error (not surprising, since the container is using overlay now):
> 
> > s6-overlay-suexec: fatal: can only run as pid 1
> 
> I used the init setting because before overlay was introduced, there was nothing to properly handle interrupt signals in the container.

Heya, 

The readme needs updating now that we've moved to S6 regardless. 

`init` has never really been recommended by us, nor does it work well with Borgmatic (as can be seen in #173). If people choose to use `init`, that's at their own discretion, but once we update the readme, we'll make sure to mention that it has S6 in, and elaborate a couple of paragraphs on what S6 does so that people are deterred from using `init` package. 

I'd be happy to review any blurb you think may be suitable to be added to the readme so that it reduces or eliminates any confusion.

Cheers!

**Pull request merged**
